### ISSUE TITLE
Provide method that runs before the Scheduler's event loop in TaskExt 

### ISSUE THREAD

**Issue by nemanja-boric-sociomantic** [sociomantic-tsunami/ocean#500]
At the moment, `DaemonApp` will start event handling and then it will schedule the `run` task
and start the scheduler:

https://github.com/sociomantic-tsunami/ocean/blob/b200e555190591538c44ca789c5dc26fc2540928/src/ocean/util/app/DaemonApp.d#L542-L545

https://github.com/sociomantic-tsunami/ocean/blob/b200e555190591538c44ca789c5dc26fc2540928/src/ocean/util/app/ext/TaskExt.d#L129

The problem with this is that, given that `processConfig` runs after the constructor, the `run` method is the only appropriate place to initialize the application's components. That leads to trouble if the `onStatsTimer` or any other event (like `SignalExt` or `UnixSocketExt`) signals and run its callback assuming the application is fully initialized while the `run` hasn't yet run or hasn't finished and had a context switch (the issue is easy to see if you run your application close to 0 or 30 seconds mark: https://github.com/sociomantic-tsunami/ocean/blob/b200e555190591538c44ca789c5dc26fc2540928/src/ocean/util/app/DaemonApp.d#L509-L512 , but any other registered event source may trigger this bug).

Requiring explicitly from the user to call `startEventHandling` is error prone, since it can be easily forgotten. The alternative could be to allow extra method to be overriden to put the initialization code, or to override `run` and call `DaemonApp`'s run after the initialization has completed.

**Comment by joseph-wakeling-sociomantic** [comment]
What I would suggest/request is that the method concerned should be possible to use with all resources that one would have prior to `this.startEventHandling` and `theScheduler.eventLoop` being called.  So e.g. one should be able to use `theScheduler.epoll` (to instantiate swarm clients), etc. etc.

**Comment by joseph-wakeling-sociomantic** [comment]
HOWEVER, I wonder if it might be worth just rolling back the change to introduce :00 and :30 stats timing.  The reason is simply that the "guarantee" it provides is not very strong: writing and/or calculating stats can take time, and it is entirely possible that this will mess up the "promise" of the fixed stats times.

**Comment by nemanja-boric-sociomantic** [comment]
That's not a robust solution because the stats timer is not the only event that you start in `startEventHandling`: also the signal extension and the unix domain socket extension are started and getting any signal or connecting any client (it could easily be the monitor application) will invoke the callbacks leading to the same problem.

**Comment by joseph-wakeling-sociomantic** [comment]
Yes, fair enough.  And even if one did roll back the changes to `onStatsTimer` scheduling, it would be good to have the guarantee of things being instantiated before event scheduling starts.

**Comment by nemanja-boric-sociomantic** [comment]
@mihails-strasuns-sociomantic raises a note that `run` is not the only appropriate place to initialize the application's components - the `processArgs` and `processConfig` are also appropriate and used. Under that assumption this ticket is actually invalid.

**Comment by joseph-wakeling-sociomantic** [comment]
> @mihails-strasuns-sociomantic raises a note that run is not the only appropriate place to initialize the application's components - the processArgs and processConfig are also appropriate and used.

Yes, I'd be happy to pursue this direction (I think I even suggested it myself in our discussion), but with one caveat: can we rely on `theScheduler` (more specifically, `theScheduler.epoll`) being initialized by the time that `process{Args,Config}` are called?

**Comment by joseph-wakeling-sociomantic** [comment]
Question: are there going to be any issues (moving forward) with swarm clients (`DhtClient`, `DmqClient`, `DlsClient`) being instantiated in `processConfig` instead of inside a task context?

**Comment by joseph-wakeling-sociomantic** [comment]
> can we rely on `theScheduler` (more specifically, `theScheduler.epoll`) being initialized by the time that `process{Args,Config}` are called?

I've tried this out.  The answer is no :-(

**Comment by mihails-strasuns-sociomantic** [comment]
See linked bug fix PR.

**Comment by mihails-strasuns** [comment]
We should think about a better support for this problem as part of https://github.com/sociomantic-tsunami/ocean/pull/544
